### PR TITLE
Bump Apache Avro to 1.9.1

### PR DESCRIPTION
**Pull request by Fokko**
Apache Avro 1.9.1 brings a lot of new features:
* Deprecate Joda-Time in favor of Java8 JSR310 and setting it as default
* Remove support for Hadoop 1.x
* Move from Jackson 1.x to 2.9
* Add ZStandard Codec
* Lots of updates on the dependencies to fix CVE's
* Remove Jackson classes from public API
* Apache Avro is built by default with Java 8
* Apache Avro is compiled and tested with Java 11 to guarantee compatibility
* Apache Avro MapReduce is compiled and tested with Hadoop 3
* Apache Avro is now leaner, multiple dependencies were removed: guava, paranamer, commons-codec, and commons-logging
* and many, many more!

### CODE REVIEW

**Comment by Fokko** [comment]
CLA has been submitted.

**Comment by martint** [comment]
@cla-bot check

File: presto-product-tests/pom.xml [hunk 1]
```diff
@@ -198,6 +198,23 @@
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
```
[the remark this answers is not in the document]
Comment (Fokko) [reply presto-product-tests/pom.xml]: Makes sense, I've reverted the SpecificSchema's. I have a strong preference towards Specific over Dynamic since the specific errors show up at compile-time, but I can understand that for Presto the generic schema is more common.

File: presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java [hunk 2]
```diff
@@ -86,10 +86,30 @@ private static String getAvroSchema(String name, String dataType)
...
 
+    private static String determineDefaultValue(String dataType)
+    {
+        // Apache Avro 1.9 got more strict on the nullability of datatypes.
+        // If the datatype is not nullable, then we don't want to set the null as default.
+        if (dataType.contains("\"array\"")) {
+            // In the case of an Array we just will return an empty array
+            return ", \"default\": []";
+        }
+        else if (dataType.contains("\"map\"")) {
```
[the remark this answers is not in the document]
Comment (Fokko) [reply presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java]: Fair, removed the else'es

File: pom.xml [hunk 3]
```diff
@@ -45,6 +45,7 @@
 
         <dep.antlr.version>4.7.1</dep.antlr.version>
         <dep.airlift.version>0.184</dep.airlift.version>
+        <dep.avro.version>1.9.1</dep.avro.version>
```
[the remark this answers is not in the document]
Comment (Fokko) [reply pom.xml]: We're still having two places where we have to set the version of Avro. Can you give an example of how to fix this?

File: presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java [hunk 4]
```diff
@@ -146,12 +166,9 @@ private static String getAvroSchema(Map<String, String> fields)
     {
         GenericData.Record record = new GenericData.Record(schema);
         values.forEach(record::put);
-        try {
-            DataFileWriter<GenericRecord> dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<>(schema));
-
+        try (DataFileWriter<GenericRecord> dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<>(schema))) {
```
[the remark this answers is not in the document]
Comment (Fokko) [reply presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java]: Done.

**Comment by Fokko** [comment]
Thank you @electrum for the review. From Presto's perspective, the impact is minimal, especially from the user's side. The format is still binary compatible, so it is able to read older versions of Avro.
From the developer perspective, there are some advantages. For example removal of several libraries that caused issues in the past, such as guava. Also, paranamer, commons-codec, and commons-logging have been removed making Avro much smaller,

File: presto-product-tests/src/main/java/io/prestosql/tests/kafka/KafkaAvroSmokeTest.java [hunk 5]
```diff
@@ -107,7 +107,9 @@ private static Requirement createAvroTable(String schemaPath, String tableName,
         ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
         GenericData.Record record = new GenericData.Record(schema);
         values.forEach(record::put);
-        try (DataFileWriter<GenericRecord> dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<>(schema))) {
+        try {
```
Comment (electrum) [inline presto-product-tests/src/main/java/io/prestosql/tests/kafka/KafkaAvroSmokeTest.java]: These lines shouldn't be changed in this commit. (looks like a rebase error)

File: presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java (hunk 4 above)
Comment (electrum) [inline presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java]: This should be fixed in the other commit.

File: presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java [hunk 6]
```diff
@@ -77,7 +77,7 @@
     private static final AvroRowDecoderFactory DECODER_FACTORY = new AvroRowDecoderFactory();
 
     private static final Metadata METADATA = createTestMetadataManager();
-    private static final Type VACHAR_MAP_TYPE = METADATA.getType(parseTypeSignature("map(varchar,varchar)"));
```
Comment (electrum) [inline presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java]: Please fix this typo in a separate commit as well.

File: pom.xml [hunk 7]
```diff
@@ -1283,6 +1283,7 @@
                                     <exclude>org.codehaus.plexus:plexus-utils</exclude>
                                     <exclude>com.google.guava:guava</exclude>
                                     <exclude>com.fasterxml.jackson.core:jackson-core</exclude>
+                                    <exclude>com.fasterxml.jackson.core:jackson-databind</exclude>
```
Comment (electrum) [inline pom.xml]: We shouldn't need this after updating to the latest Airbase (yet to be released) that updates Jackson.

I'm working on releasing it now and will put up a separate PR to update.

File: presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java (hunk 4 above)
Comment (Fokko) [reply presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java]: This is in a separate commit: https://github.com/prestosql/presto/pull/1602/commits/9763db5048ccddeec1a73efc706709ee4ef24700

Or do you want to have it in a separate PR?

Comment (Fokko) [reply presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java]: Done

File: presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java (hunk 6 above)
Comment (Fokko) [reply presto-record-decoder/src/test/java/io/prestosql/decoder/avro/TestAvroDecoder.java]: Done

File: presto-product-tests/src/main/java/io/prestosql/tests/kafka/KafkaAvroSmokeTest.java (hunk 5 above)
Comment (Fokko) [reply presto-product-tests/src/main/java/io/prestosql/tests/kafka/KafkaAvroSmokeTest.java]: Something looks off indeed

**Comment by Fokko** [comment]
I've fixed the issues and split up the work in separate commits.

**Comment by Fokko** [comment]
Fixed merge conflict.

**Comment by Fokko** [comment]
@electrum Any outstanding questions on the PR?

**Comment by electrum** [comment]
Merged, thanks!

**Comment by electrum** [comment]
Merged as 6059a07d5e0ed595e231bf89c726f94e7a985d89 with a couple of minor tweaks
